Got Error in Windows while executing PyWBEM

I'm new to PyWBEM and got this issue while executing this script

(tools) D:\Soft\PyWBEM\pywbemtools\pywbemtools>workon tools
(tools) D:\Soft\PyWBEM\pywbemtools\pywbemtools>make develop
  File "<string>", line 1
    import sys; sys.stdout.write('s.sys.version_info[0:3])
                                                         ^
SyntaxError: EOL while scanning string literal
  File "<string>", line 1
    import sys; sys.stdout.write('sys.version_info[0])
                                                     ^
SyntaxError: EOL while scanning string literal
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: 'str' object is not callable
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
wget -q http://pywbem.readthedocs.io/en/latest/_downloads/pywbem_os_setup.sh
process_begin: CreateProcess((null), wget -q http://pywbem.readthedocs.io/en/latest/_downloads/pywbem_os_setup.sh, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [pywbem_os_setup.sh] Error 2

(tools) D:\Soft\PyWBEM\pywbemtools\pywbemtools>https://github.com/pywbem/pywbemtools.githttps://github.com/pywbem/pywbemtools.git

Hi Andreas,

Again getting the same error message. I tried even creating new virtual environment but it is throwing the same error message while executing the make develop command.


(tools) D:\Soft\PyWBEM\pywbemtools\pywbemtools>mkvirtualenv raam
Using base prefix 'c:\\users\\raamesh-1460\\appdata\\local\\programs\\python\\python37-32'
New python executable in C:\Users\raamesh-1460\Envs\raam\Scripts\python.exe
Installing setuptools, pip, wheel...
done.

(raam) D:\Soft\PyWBEM\pywbemtools\pywbemtools>workon

Pass a name to activate one of the following virtualenvs:
==============================================================================
PyWBEM
pywbemdev
raam
TestEnv
tools

(raam) D:\Soft\PyWBEM\pywbemtools\pywbemtools>workon raam
(raam) D:\Soft\PyWBEM\pywbemtools\pywbemtools>make develop
  File "<string>", line 1
    import sys; sys.stdout.write('s.sys.version_info[0:3])
                                                         ^
SyntaxError: EOL while scanning string literal
  File "<string>", line 1
    import sys; sys.stdout.write('sys.version_info[0])
                                                     ^
SyntaxError: EOL while scanning string literal
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: 'str' object is not callable
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
  File "<string>", line 1
    $(printf 'try:\n from pbr.version import VersionInfo\nexcept ImportError:\n pass\nelse:\n print(VersionInfo(\042pywbemtools\042).release_string())\n')
    ^
SyntaxError: invalid syntax
wget -q http://pywbem.readthedocs.io/en/latest/_downloads/pywbem_os_setup.sh
process_begin: CreateProcess((null), wget -q http://pywbem.readthedocs.io/en/latest/_downloads/pywbem_os_setup.sh, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [pywbem_os_setup.sh] Error 2

Hi Andreas,
Thanks for the update. Checked out the branch.

Now, I'm getting this error

(Testing) D:\Soft\Raam\pywbem>workon Testing
(Testing) D:\Soft\Raam\pywbem>make develop
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'pbr'
python -m pip install --upgrade pip
Requirement already up-to-date: pip in
c:\users\raamesh-1460\envs\testing\lib\site-packages (19.0.1)
touch pip_upgrade.done
process_begin: CreateProcess((null), touch pip_upgrade.done, ...) failed.
make (e=2): The system cannot find the file specified.
make: *** [pip_upgrade.done] Error 2




On Thu, 7 Feb 2019 at 13:01, Andreas Maier <notifications@github.com> wrote:

> Raamesh,
> In the update to PR #1636 <https://github.com/pywbem/pywbem/pull/1636>
> from 2 days ago, I moved the inline Python in the makefile into small
> scripts. So an error like:
>
> raam) D:\Soft\PyWBEM\pywbemtools\pywbemtools>make develop
> File "", line 1
> import sys; sys.stdout.write('s.sys.version_info[0:3])
>
> shows that you did not use the latest version of the branch of that PR.
> The update to the PR may have crossed your attempt to use it.
>
> Could you please refresh your local version of the PR (e.g. git pull when
> branch andy/make-windows-no-quotes is checked out) and retry?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pywbem/pywbem/issues/1631#issuecomment-461314858>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUrJzH_lCJZyhNLG-OhmCI0gFqYev4C5ks5vK9ZMgaJpZM4aRlmT>
> .
>


-- 
Regards
Raamesh Keerthi N J

Today is your day, your mountain is waiting, so get on your own way...

--- a/.travis.yml
+++ b/.travis.yml
@@ -209,6 +209,7 @@ install:
     fi
 # End of m2crypto quick fix.
   - pip list
+  - make platform env
   - make install
   - pip list
   - make develop

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,36 +5,86 @@ environment:
 # Python 2.6 is not supported by M2Crypto on Windows (nor by the earlier used
 # M2CryptoWin32/Win64; its __m2crypto.pyd file depends on python27.dll).
 #------------------------
+
     - PYTHON_VERSION: 2.7
       PYTHON_ARCH: 32
       PYTHON_HOME: C:\Python27
+      UNIX_PATH: none
+
+#    - PYTHON_VERSION: 2.7
+#      PYTHON_ARCH: 32
+#      PYTHON_HOME: C:\Python27
+#      UNIX_PATH: C:\cygwin\bin
+
 #    - PYTHON_VERSION: 2.7
 #      PYTHON_ARCH: 64
 #      PYTHON_HOME: C:\Python27-x64
+#      UNIX_PATH: none
+
+#    - PYTHON_VERSION: 2.7
+#      PYTHON_ARCH: 64
+#      PYTHON_HOME: C:\Python27-x64
+#      UNIX_PATH: C:\cygwin64\bin
+
+    - PYTHON_VERSION: 2.7
+      PYTHON_ARCH: 64
+      PYTHON_HOME: C:\Python27-x64
+      UNIX_PATH: C:\msys64\usr\bin
+
 #    - PYTHON_VERSION: 3.4
 #      PYTHON_ARCH: 32
 #      PYTHON_HOME: C:\Python34
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.4
 #      PYTHON_ARCH: 64
 #      PYTHON_HOME: C:\Python34-x64
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.5
 #      PYTHON_ARCH: 32
 #      PYTHON_HOME: C:\Python35
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.5
 #      PYTHON_ARCH: 64
 #      PYTHON_HOME: C:\Python35-x64
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.6
 #      PYTHON_ARCH: 32
 #      PYTHON_HOME: C:\Python36
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.6
 #      PYTHON_ARCH: 64
 #      PYTHON_HOME: C:\Python36-x64
+#      UNIX_PATH: none
+
 #    - PYTHON_VERSION: 3.7
 #      PYTHON_ARCH: 32
 #      PYTHON_HOME: C:\Python37
+#      UNIX_PATH: none
+
+#    - PYTHON_VERSION: 3.7
+#      PYTHON_ARCH: 32
+#      PYTHON_HOME: C:\Python37
+#      UNIX_PATH: C:\cygwin\bin
+
+#    - PYTHON_VERSION: 3.7
+#      PYTHON_ARCH: 64
+#      PYTHON_HOME: C:\Python37-x64
+#      UNIX_PATH: none
+
     - PYTHON_VERSION: 3.7
       PYTHON_ARCH: 64
       PYTHON_HOME: C:\Python37-x64
+      UNIX_PATH: C:\cygwin64\bin
+
+#    - PYTHON_VERSION: 3.7
+#      PYTHON_ARCH: 64
+#      PYTHON_HOME: C:\Python37-x64
+#      UNIX_PATH: C:\msys64\usr\bin
 
 configuration:
 # These values will become the values of the PACKAGE_LEVEL env.var.
@@ -81,25 +131,18 @@ install:
   - 'set $newpath=%$newpath:^^=%'
   - 'set PATH=%$newpath%'
 
-  # Add CygWin
-  # This must be after removal of Python 2.7 (because CygWin has the Unix find) and
-  # before adding Python (because CygWin also has a python).
-  - set PATH=C:\cygwin\bin;%PATH%
+  # Add UNIX path, if set in the build matrix
+  # This must be after removal of Python 2.7 (because that uses the Windows
+  # find and the UNIX env likely has the Unix find) and before adding Python
+  # (because the UNIX env may also have a python).
+  - 'if not "%UNIX_PATH%"=="none" set PATH=%UNIX_PATH%;%PATH%'
 
-  # Add Python
+  # Add Python (includes Pip)
   - reg ADD HKCU\Software\Python\PythonCore\%PYTHON_VERSION%\InstallPath /ve /d "%PYTHON_HOME%" /t REG_SZ /f
   - reg ADD HKLM\Software\Python\PythonCore\%PYTHON_VERSION%\InstallPath /ve /d "%PYTHON_HOME%" /t REG_SZ /f
-  - set PATH=%PYTHON_HOME%;%PYTHON_HOME%\Scripts;%PATH%
+  - 'set PATH=%PYTHON_HOME%;%PYTHON_HOME%\Scripts;%PATH%'
   - where python
   - python --version
-
-  ## Install InnoSetup - disabled because it is not needed
-  #- choco install -y InnoSetup
-  #- set PATH="C:\Program Files (x86)\Inno Setup 5";%PATH%
-
-  ## Install pip - disabled because pip is already installed
-  #- ps: (new-object System.Net.WebClient).Downloadfile('https://bootstrap.pypa.io/get-pip.py', 'C:\Users\appveyor\get-pip.py')
-  #- ps: Start-Process -FilePath "C:\Python27\python.exe" -ArgumentList "C:\Users\appveyor\get-pip.py" -Wait -Passthru
   - where pip
   - pip --version
 
@@ -107,60 +150,66 @@ install:
   - where choco
   - choco --version
 
+  # Install GNU make if no UNIX env is set
+  - if "%UNIX_PATH%"=="none" choco install -y make
+  - where make
+  - make --version
+
   # Circumvention:
   # The installation in pywbem_os_setup.bat installs swig but without generating
   # the swig.exe shim file (using GenShim).
   # See GenShim issue https://github.com/chocolatey/shimgen/issues/43
   # Therefore, we install swig here in order to create the swig.exe shim file.
   # TODO: Remove the circumvention once it works in pywbem_os_setup.bat.
-  - choco install -r -y --allow-empty-checksums swig
-  - where swig
+  #- choco install -y --allow-empty-checksums swig
+  #- where swig
+  #- swig -version
 
   # Examine the final environment
-  - 'echo "%PATH%"'
-  - 'echo "%PYTHONPATH%"'
-  - 'echo "%INCLUDE%"'
-  - 'echo "%LIB%"'
-  - 'echo "%SWIG_FEATURES%"'
+  - 'echo: %PATH%'
+  - 'echo: %PYTHONPATH%'
+  - 'echo: %INCLUDE%'
+  - 'echo: %LIB%'
+  - 'echo: %SWIG_FEATURES%'
 
-  # Install tox
-  - pip install tox==2.0.0
-  - where tox
-  - tox --version
+  # Verify that the commands used in pywbem_os_setup.bat are available
+  # Already verified further up: choco, pip, python.
+  - where chmod
+  - chmod --version
+  - where tar
+  - tar --version
 
-  # Verify that the commands used in tox.ini are available
-  # Already verified further up: tox, pip, python.
-  - where make
-  - make --version
+  # Install Python virtualenv and OS commands
+  #
+  # Note: Installing Swig in pywbem_os_setup.bat does not work when it runs
+  # inside of make which runs inside of tox, because it does not generate the
+  # swig.exe shim file (using GenShim). See GenShim issue
+  # https://github.com/chocolatey/shimgen/issues/43.
+  # Therefore, we invoke pywbem_os_setup.bat outside of make.
+  - pip install virtualenv
+  - virtualenv myenv
+  - myenv\Scripts\activate
+  - where python
+  - python --version
+  - where pip
+  - pip --version
+  - call pywbem_os_setup.bat install
+  - call pywbem_os_setup.bat develop
 
   # Verify that the commands used in makefile are available
   # Already verified further up: pip, python.
-  - where sh
-  - sh --version
-  - where bash
-  - bash --version
-  - where rm
-  - rm --version
-  - where mv
-  - mv --version
-  - where xargs
-  - xargs --version
-  - where grep
-  - grep --version
-  - where sed
-  - sed --version
-  - where tar
-  - tar --version
-  - where find
-  - find --version
-
-  # Verify that the commands used in pywbem_os_setup.bat are available
-  - where curl
-  - curl --version
 
   # Verify that the commands used in the pywbem testsuite are available
   - where xmllint
   - xmllint --version
+
+  # Install Python packages
+  - make platform env
+  - pip list
+  - make install
+  - pip list
+  - make develop
+  - pip list
 
 # Not a C# project, build stuff at the test step instead.
 build: false
@@ -168,4 +217,4 @@ build: false
 before_test:
 
 test_script:
-  - tox -e pywin
+  - make test

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -151,7 +151,7 @@ install:
   - choco --version
 
   # Install GNU make if no UNIX env is set
-  - if "%UNIX_PATH%"=="none" choco install -y make
+  - if "%UNIX_PATH%"=="none" tools\retry choco install -y make
   - where make
   - make --version
 

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -259,6 +259,9 @@ This version contains all fixes up to pywbem 0.12.4.
   contribute to the same repository (the original version copied objects
   from the compiler repository to the mocker repository).
 
+* Test: Fixed a bytes/unicode error in validate.py that occurred on Python 3
+  when xmllint failed validating the DTD.
+
 **Enhancements:**
 
 * Extend pywbem MOF compiler to search for dependent classes including:

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -616,6 +616,25 @@ This version contains all fixes up to pywbem 0.12.4.
 
 * Added support for running pylint also on Python 3.x. See issue #1640.
 
+* Improved the makefile for use on native Windows. See issue #1631. Details:
+
+  - Some GNU make versions on native Windows have an issue with double
+    quotes in make $(shell ..) commands; removed the use of double quotes.
+    As a result, most inline python commands have been moved into new small
+    scripts in the tools directory.
+    Also, several make targets that used to produce log files,
+    no longer can do that and the user needs to redirect the make invocation
+    in order to get a log file.
+
+  - Removed dependencies on most UNIX-like commands (touch, tee, bash, rm,
+    find, xargs) when using make on native Windows.
+
+  - Encapsulated file removal and copy commands to be compatible between
+    native Windows and other platforms.
+
+  - Updated the appveyor.yml file to check only the new, smaller, list of
+    commands.
+
 **Known issues:**
 
 * See `list of open issues`_.

--- a/docs/intro.rst
+++ b/docs/intro.rst
@@ -253,13 +253,14 @@ Installing to OS-X
 Installing to native Windows
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-This section describes the installation of pywbem into a native Windows
-environment (i.e. without using a UNIX-like environment; for that, see
-:ref:`Installing to a UNIX-like environment under Windows`):
+This section describes the installation of pywbem into a *native Windows
+environment*. That is an environment where the Windows command processor is
+used to run any commands. Tools from UNIX-like environments may or may not be
+present in the PATH.
 
 * Prerequisites:
 
-  - The Python environment into which you want to install must be the current
+  - The Python environment into which you want to install must be the active
     Python environment, and must have at least the following Python packages
     installed:
 
@@ -267,9 +268,6 @@ environment (i.e. without using a UNIX-like environment; for that, see
     - wheel
     - pip
     - pbr
-
-  - The `Chocolatey package manager`_ must be installed and its "choco" command
-    available in the PATH.
 
   - Windows command prompt in administrator mode.
 
@@ -280,30 +278,50 @@ environment (i.e. without using a UNIX-like environment; for that, see
   - `pywbem_os_setup.bat <https://raw.githubusercontent.com/pywbem/pywbem/master/pywbem_os_setup.bat>`_
     on the `master` branch of the pywbem Git repository
 
-  That script installs OS-level prerequisite packages needed by pywbem.
-
-* Execute the ``pywbem_os_setup.bat`` script in an administrator command prompt:
+* Execute the ``pywbem_os_setup.bat`` script in a Windows command prompt in
+  administrator mode:
 
   .. code-block:: bash
 
-      > pywbem_os_setup.bat
+     > pywbem_os_setup.bat
 
-  The script uses the `Chocolatey package manager`_.
+  This script checks whether the commands needed for installing and using
+  pywbem are available in the PATH. If not, it installs them via the
+  `Chocolatey package manager`_.
+
+  The following commands are needed for installing and using pywbem:
+
+  * ``swig``
+  * ``curl``
+  * ``grep``
+  * ``chmod``
+  * ``tar``
+
+  These commands can be made available in the PATH via a UNIX-like environment
+  such as `CygWin`_, `MSYS2`_, Babun, or Gow. If they are not all available in
+  the PATH, the `Chocolatey package manager`_ must be installed and its
+  ``choco`` command must be available in the PATH.
 
   This script will also install the ``M2Crypto`` Python package into the active
-  Python environment.
+  Python environment, so it must be run with the desired Python environment
+  active.
 
-  This script will also install Win32 OpenSSL from
+  This script will also download and install Win32 OpenSSL from
   https://slproweb.com/products/Win32OpenSSL.html.
 
 * Install pywbem (and its other prerequisite Python packages) into the active
-  Python environment:
+  Python environment by running in a Windows command prompt in administrator
+  mode:
 
   .. code-block:: bash
 
-      > pip install pywbem
+     > pip install pywbem
 
 .. _`Chocolatey package manager`: https://chocolatey.org/
+
+.. _`CygWin`: https://cygwin.org/
+
+.. _`MSYS2`: https://www.msys2.org/
 
 
 .. _`Installing to a UNIX-like environment under Windows`:
@@ -311,8 +329,10 @@ environment (i.e. without using a UNIX-like environment; for that, see
 Installing to a UNIX-like environment under Windows
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-This section describes the installation of pywbem into a UNIX-like environment
-under Windows (such as CygWin, MinGW, Babun, or Gow).
+This section describes the installation of pywbem into a *UNIX-like environment
+under Windows* (such as `CygWin`_, `MSYS2`_, Babun, or Gow). That is an
+environment where the UNIX/Linux shell of the UNIX-like environment (such as
+``bash`` or ``sh``) is used to run any commands.
 
 Note that Unix-like environments on Windows bring their own Python, so in such
 an environment, you install into that Python, and not into the Python of
@@ -341,11 +361,12 @@ Windows.
     tell us by `opening an issue <https://github.com/pywbem/pywbem/issues>`_).
 
 * Install pywbem (and its other prerequisite Python packages) into the active
-  Python environment:
+  Python environment, by running in the UNIX/Linux shell of the UNIX-like
+  environment:
 
   .. code-block:: bash
 
-      > pip install pywbem
+      $ pip install pywbem
 
   If the Swig compilation during installation of ``M2Crypto`` fails, there may
   be components of the UNIX-like environment missing (see first item).

--- a/makefile
+++ b/makefile
@@ -4,24 +4,32 @@
 # Supported OS platforms for this makefile:
 #     Linux (any distro)
 #     OS-X
-#     Windows with UNIX-like env such as CygWin (with Python in UNIX-like env)
-#     native Windows (with Python in Windows)
+#     Windows with UNIX-like env such as CygWin (with a UNIX-like shell and
+#       Python in the UNIX-like env)
+#     native Windows (with the native Windows command processor and Python in
+#       Windows)
 #
 # Prerequisites for running this makefile:
 #   These commands are used on all supported OS platforms. On native Windows,
-#   they may be provided by CygWin:
+#   they may be provided by UNIX-like environments such as CygWin:
 #     make (GNU make)
-#     bash
-#     echo, rm, mv, find, xargs, tee, touch, chmod, wget
 #     python (This Makefile uses the active Python environment, virtual Python
-#        environments are supported)
+#       environments are supported)
 #     pip (in the active Python environment)
 #     twine (in the active Python environment)
-#   These additional commands are used on Linux, OS-X and Windows with UNIX env:
+#   These additional commands are used on Linux, OS-X and on Windows with
+#   UNIX-like environments:
 #     uname
+#     rm, find, xargs, cp
+#     The commands listed in pywbem_os_setup.sh
 #   These additional commands are used on native Windows:
-#     cmd
+#     del, copy, rmdir
+#     The commands listed in pywbem_os_setup.bat
 # ------------------------------------------------------------------------------
+
+# No built-in rules needed:
+MAKEFLAGS += --no-builtin-rules
+.SUFFIXES:
 
 # Python / Pip commands
 ifndef PYTHON_CMD
@@ -45,31 +53,80 @@ else
   endif
 endif
 
+# Make variables are case sensitive and some native Windows environments have
+# ComSpec set instead of COMSPEC.
+ifndef COMSPEC
+  ifdef ComSpec
+    COMSPEC = $(ComSpec)
+  endif
+endif
+
 # Determine OS platform make runs on.
+#
+# The PLATFORM variable is set to one of:
+# * Windows_native: Windows native environment (the Windows command processor
+#   is used as shell and its internal commands are used, such as "del").
+# * Windows_UNIX: A UNIX-like envieonment on Windows (the UNIX shell and its
+#   internal commands are used, such as "rm").
+# * Linux: Some Linux distribution
+# * Darwin: OS-X / macOS
+#
+# This in turn determines the type of shell that is used by make when invoking
+# commands, and the set of internal shell commands that is assumed to be
+# available (e.g. "del" for the Windows native command processor and "rm" for
+# a UNIX-like shell). Note that GNU make always uses the value of the SHELL
+# make variable to invoke the shell for its commands, but it does not always
+# read that variable from the environment. In fact, the approach GNU make uses
+# to set the SHELL make variable is very special, see
+# https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html.
+# On native Windows this seems to be implemented differently than described:
+# SHELL is not set to COMSPEC, so we do that here.
+#
 # Note: Native Windows and CygWin are hard to distinguish: The native Windows
-# envvars are set in CygWin as well. Using uname will display CYGWIN_NT-.. on
-# both platforms. If the CygWin make is used on native Windows, most of the
-# CygWin behavior is then visible in context of that make (e.g. a SHELL envvar
-# is set, the PATH envvar gets converted to UNIX syntax, execution of batch
-# files requires execute permission, etc.). The check below with
-# :/usr/local/bin: being in PATH was found to work even when using the CygWin
-# make on native Windows.
+# envvars are set in CygWin as well. COMSPEC (or ComSpec) is set on both
+# platforms. Using "uname" will display CYGWIN_NT-.. on both platforms. If the
+# CygWin make is used on native Windows, most of the CygWin behavior is visible
+# in context of that make (e.g. a SHELL variable is set, PATH gets converted to
+# UNIX syntax, execution of batch files requires execute permission, etc.).
 ifeq ($(OS),Windows_NT)
-  ifeq ($(findstring :/usr/local/bin:,$(PATH)),:/usr/local/bin:)
-    PLATFORM := CygWin
+  ifdef PWD
+    PLATFORM := Windows_UNIX
   else
-    PLATFORM := Windows
+    PLATFORM := Windows_native
+    ifdef COMSPEC
+      SHELL := $(subst \,/,$(COMSPEC))
+    else
+      SHELL := cmd.exe
+    endif
+    .SHELLFLAGS := /c
   endif
 else
   # Values: Linux, Darwin
   PLATFORM := $(shell uname -s)
 endif
 
-ifeq ($(PLATFORM),Windows)
-  # Using the CygWin find
-  FIND := /bin/find
+ifeq ($(PLATFORM),Windows_native)
+  # Note: The substituted backslashes must be doubled.
+  # Remove files (blank-separated list of wildcard path specs)
+  RM_FUNC = del /f /q $(subst /,\\,$(1))
+  # Remove files recursively (single wildcard path spec)
+  RM_R_FUNC = del /f /q /s $(subst /,\\,$(1))
+  # Remove directories (blank-separated list of wildcard path specs)
+  RMDIR_FUNC = rmdir /q /s $(subst /,\\,$(1))
+  # Remove directories recursively (single wildcard path spec)
+  RMDIR_R_FUNC = rmdir /q /s $(subst /,\\,$(1))
+  # Copy a file, preserving the modified date
+  CP_FUNC = copy /y $(subst /,\\,$(1)) $(subst /,\\,$(2))
+  ENV = set
+  WHICH = where
 else
-  FIND := find
+  RM_FUNC = rm -f $(1)
+  RM_R_FUNC = find . -type f -name '$(1)' -delete
+  RMDIR_FUNC = rm -rf $(1)
+  RMDIR_R_FUNC = find . -type d -name '$(1)' | xargs -n 1 rm -rf
+  CP_FUNC = cp -r $(1) $(2)
+  ENV = env | sort
+  WHICH = which
 endif
 
 # Name of this Python package
@@ -95,12 +152,12 @@ coverage_html_dir := coverage_html
 # e.g. because the pywbem.egg-info directory or the PKG-INFO file are deleted,
 # when a new version tag has been assigned. Therefore, this variable is assigned with
 # "=" so that it is evaluated every time it is used.
-package_version = $(shell $(PYTHON_CMD) -c "from pbr.version import VersionInfo; print(VersionInfo('$(package_name)').release_string())")
+package_version = $(shell $(PYTHON_CMD) tools/package_version.py $(package_name))
 
 # Python versions
-python_version := $(shell $(PYTHON_CMD) -c "import sys; sys.stdout.write('{0}.{1}.{2}'.format(*sys.version_info[0:3]))")
-python_mn_version := $(shell $(PYTHON_CMD) -c "import sys; sys.stdout.write('{0}{1}'.format(*sys.version_info[0:2]))")
-python_m_version := $(shell $(PYTHON_CMD) -c "import sys; sys.stdout.write('{0}'.format(sys.version_info[0:1]))")
+python_version := $(shell $(PYTHON_CMD) tools/python_version.py 3)
+python_mn_version := $(shell $(PYTHON_CMD) tools/python_version.py 2)
+python_m_version := $(shell $(PYTHON_CMD) tools/python_version.py 1)
 
 # Directory for the generated distribution files
 dist_dir := dist
@@ -202,24 +259,24 @@ test_yaml_files := \
     $(wildcard tests/unittest/*/*.y*ml) \
     $(wildcard tests/functiontest/*.y*ml) \
 
-# Test log
-test_log_file := test_$(python_mn_version).log
-test_tmp_file := test_$(python_mn_version).tmp.log
-test_end2end_log_file := test_end2end_$(python_mn_version).log
-test_end2end_tmp_file := test_end2end_$(python_mn_version).tmp.log
-
 ifdef TESTCASES
-pytest_opts := $(TESTOPTS) -k $(TESTCASES)
+  pytest_opts := $(TESTOPTS) -k $(TESTCASES)
 else
-pytest_opts := $(TESTOPTS)
+  pytest_opts := $(TESTOPTS)
 endif
 pytest_end2end_opts := -v --tb=short $(pytest_opts)
 
-pytest_warnings := default
 ifeq ($(python_m_version),3)
-  pytest_end2end_warnings := default,ignore::DeprecationWarning,ignore::PendingDeprecationWarning,ignore::ResourceWarning
+  pytest_warnings := --pythonwarnings=default
+  pytest_end2end_warnings_opts := --pythonwarnings=default,ignore::DeprecationWarning,ignore::PendingDeprecationWarning,ignore::ResourceWarning
 else
-  pytest_end2end_warnings := default,ignore::DeprecationWarning,ignore::PendingDeprecationWarning
+  ifeq ($(python_mn_version),2.6)
+    pytest_warnings :=
+    pytest_end2end_warnings_opts :=
+  else
+    pytest_warnings := --pythonwarnings=default
+    pytest_end2end_warnings_opts := --pythonwarnings=default,ignore::DeprecationWarning,ignore::PendingDeprecationWarning
+  endif
 endif
 
 # Files to be put into distribution archive.
@@ -244,31 +301,34 @@ dist_dependent_files := \
     $(wildcard $(package_name)/*.py) \
     $(wildcard $(mock_package_name)/*.py) \
 
-# No built-in rules needed:
-.SUFFIXES:
-
 .PHONY: help
 help:
-	@echo "Makefile for $(package_name) repository of pywbem project"
-	@echo "Package version will be: $(package_version)"
-	@echo "Uses the currently active Python environment: Python $(python_version)"
+	@echo "Makefile for $(package_name) package"
 	@echo "Platform: $(PLATFORM)"
+	@echo "Shell used for commands: $(SHELL)"
+	@echo "Shell flags: $(.SHELLFLAGS)"
+	@echo "Make version: $(MAKE_VERSION)"
+	@echo "Python location: $(shell $(WHICH) python)"
+	@echo "Python version: $(python_version)"
+	@echo "$(package_name) package version: $(package_version)"
 	@echo ""
 	@echo "Make targets:"
 	@echo "  install    - Install pywbem and its Python installation and runtime prereqs (includes install_os once after clobber)"
 	@echo "  develop    - Install Python development prereqs (includes develop_os once after clobber)"
 	@echo "  build      - Build the distribution archive files in: $(dist_dir)"
 	@echo "  builddoc   - Build documentation in: $(doc_build_dir)"
-	@echo "  check      - Run Flake8 on sources and save results in: flake8.log"
-	@echo "  pylint     - Run PyLint on sources and save results in: pylint.log"
-	@echo "  test       - Run unit and function tests and save results in: $(test_log_file)"
+	@echo "  check      - Run Flake8 on sources"
+	@echo "  pylint     - Run PyLint on sources"
+	@echo "  test       - Run unit and function tests"
 	@echo "  all        - Do all of the above"
-	@echo "  end2end    - Run end2end tests and save results in: $(test_end2end_log_file)"
+	@echo "  end2end    - Run end2end tests"
 	@echo "  install_os - Install OS-level installation and runtime prereqs"
 	@echo "  develop_os - Install OS-level development prereqs"
 	@echo "  upload     - build + upload the distribution archive files to PyPI"
 	@echo "  clean      - Remove any temporary files"
 	@echo "  clobber    - Remove everything created to ensure clean start - use after setting git tag"
+	@echo "  platform   - Display the information about the platform as seen by make"
+	@echo "  env        - Display the environment as seen by make"
 	@echo ""
 	@echo "Environment variables:"
 	@echo "  COVERAGE_REPORT - When set, the 'test' target creates a coverage report as"
@@ -294,72 +354,86 @@ help:
 	@echo "  PIP_CMD - Pip command to be used. Useful for Python 3 in some envs."
 	@echo "      Optional, defaults to 'pip'."
 
+.PHONY: platform
+platform:
+	@echo "Platform: $(PLATFORM)"
+	@echo "Shell used for commands: $(SHELL)"
+	@echo "Shell flags: $(.SHELLFLAGS)"
+	@echo "Make version: $(MAKE_VERSION)"
+	@echo "Python location: $(shell $(WHICH) python)"
+	@echo "Python version: $(python_version)"
+	@echo "$(package_name) package version: $(package_version)"
+
+.PHONY: env
+env:
+	@echo "Environment as seen by make:"
+	$(ENV)
+
 .PHONY: _check_version
 _check_version:
 ifeq (,$(package_version))
-	@echo 'Error: Package version could not be determined (requires pbr; run "make install")'
-	@false
-else
-	@true
+	$(error Package version could not be determined - requires pbr - run "make install")
 endif
 
-pip_upgrade.done:
-ifeq ($(python_mn_version),26)
+pip_upgrade.done: makefile
+	-$(call RM_FUNC,$@)
+ifeq ($(python_mn_version),2.6)
 	$(PIP_CMD) install $(pip_level_opts) pip
 else
 	$(PYTHON_CMD) -m pip install $(pip_level_opts) pip
 endif
-	touch pip_upgrade.done
+	echo "done" >$@
 
-install_basic.done: pip_upgrade.done
+install_basic.done: makefile pip_upgrade.done
 	@echo "makefile: Installing/upgrading basic Python packages with PACKAGE_LEVEL=$(PACKAGE_LEVEL)"
-ifeq ($(python_mn_version),26)
+	-$(call RM_FUNC,$@)
+ifeq ($(python_mn_version),2.6)
 	$(PIP_CMD) install importlib
 endif
 	$(PYTHON_CMD) remove_duplicate_setuptools.py
 # Keep the condition for the 'wheel' package consistent with the requirements & constraints files.
 # The approach with "python -m pip" is needed for Windows because pip.exe may be locked,
 # but it is not supported on Python 2.6 (which is not supported with pywbem on Windows).
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	$(PIP_CMD) install $(pip_level_opts) setuptools 'wheel<0.30.0'
 else
 	$(PIP_CMD) install $(pip_level_opts) setuptools wheel
 endif
 	$(PIP_CMD) install $(pip_level_opts) pbr
-	touch install_basic.done
+	echo "done" >$@
 	@echo "makefile: Done installing/upgrading basic Python packages"
 
 .PHONY: install_os
 install_os: install_os.done
 	@echo "makefile: Target $@ done."
 
-install_os.done: pip_upgrade.done pywbem_os_setup.sh pywbem_os_setup.bat
+install_os.done: makefile pip_upgrade.done pywbem_os_setup.sh pywbem_os_setup.bat
 	@echo "makefile: Installing OS-level installation and runtime requirements"
 	@echo "Debug: PATH=$(PATH)"
-ifeq ($(PLATFORM),Windows)
-	cmd /d /c pywbem_os_setup.bat install
+	-$(call RM_FUNC,$@)
+ifeq ($(PLATFORM),Windows_native)
+	pywbem_os_setup.bat install
 else
 	./pywbem_os_setup.sh install
 endif
-	touch install_os.done
+	echo "done" >$@
 	@echo "makefile: Done installing OS-level installation and runtime requirements"
 
 .PHONY: _show_bitsize
 _show_bitsize:
 	@echo "makefile: Determining bit size of Python executable"
-	$(PYTHON_CMD) -c "import ctypes; print(ctypes.sizeof(ctypes.c_void_p)*8)"
-	$(PYTHON_CMD) -c "import sys; print(64 if sys.maxsize > 2**32 else 32)"
-	$(PYTHON_CMD) -c "import platform; print(int(platform.architecture()[0].rstrip('bit')))"
+	$(PYTHON_CMD) tools/python_bitsize.py
 	@echo "makefile: Done determining bit size of Python executable"
 
-install_pywbem.done: pip_upgrade.done requirements.txt setup.py setup.cfg
+install_pywbem.done: makefile pip_upgrade.done requirements.txt setup.py setup.cfg
 	@echo "makefile: Installing pywbem (editable) and its Python runtime prerequisites (with PACKAGE_LEVEL=$(PACKAGE_LEVEL))"
-	rm -Rf build $(package_name).egg-info .eggs
-	rm -f PKG-INFO
+	-$(call RM_FUNC,$@)
+	-$(call RM_FUNC,PKG-INFO)
+	-$(call RMDIR_FUNC,build $(package_name).egg-info .eggs)
 	$(PIP_CMD) install $(pip_level_opts) -r requirements.txt
 	$(PIP_CMD) install $(pip_level_opts) -e .
-	cp -r $(package_name).egg-info/PKG-INFO .
-	touch install_pywbem.done
+	$(call CP_FUNC,$(package_name).egg-info/PKG-INFO,.)
+	echo "done" >$@
 	@echo "makefile: Done installing pywbem and its Python runtime prerequisites"
 
 .PHONY: install
@@ -367,9 +441,10 @@ install: install.done
 	@echo "makefile: Target $@ done."
 
 install.done: makefile install_os.done install_basic.done install_pywbem.done
-	$(PYTHON_CMD) -c "import $(package_name); print('ok, version={0}'.format($(package_name).__version__))"
-	$(PYTHON_CMD) -c "import $(mock_package_name); print('ok')"
-	touch install.done
+	-$(call RM_FUNC,$@)
+	$(PYTHON_CMD) -c "import $(package_name)"
+	$(PYTHON_CMD) -c "import $(mock_package_name)"
+	echo "done" >$@
 
 .PHONY: develop_os
 develop_os: develop_os.done
@@ -377,12 +452,13 @@ develop_os: develop_os.done
 
 develop_os.done: pywbem_os_setup.sh
 	@echo "makefile: Installing OS-level development requirements"
-ifeq ($(PLATFORM),Windows)
-	cmd /d /c pywbem_os_setup.bat develop
+	-$(call RM_FUNC,$@)
+ifeq ($(PLATFORM),Windows_native)
+	pywbem_os_setup.bat develop
 else
 	./pywbem_os_setup.sh develop
 endif
-	touch develop_os.done
+	echo "done" >$@
 	@echo "makefile: Done installing OS-level development requirements"
 
 .PHONY: develop
@@ -391,8 +467,9 @@ develop: develop.done
 
 develop.done: pip_upgrade.done install.done develop_os.done install_basic.done dev-requirements.txt
 	@echo "makefile: Installing Python development requirements (with PACKAGE_LEVEL=$(PACKAGE_LEVEL))"
+	-$(call RM_FUNC,$@)
 	$(PIP_CMD) install $(pip_level_opts) -r dev-requirements.txt
-	touch develop.done
+	echo "done" >$@
 	@echo "makefile: Done installing Python development requirements"
 
 .PHONY: build
@@ -404,11 +481,11 @@ builddoc: html
 	@echo "makefile: Target $@ done."
 
 .PHONY: check
-check: flake8.log safety.log
+check: flake8.done safety.done
 	@echo "makefile: Target $@ done."
 
 .PHONY: pylint
-pylint: pylint.log
+pylint: pylint.done
 	@echo "makefile: Target $@ done."
 
 .PHONY: all
@@ -418,8 +495,8 @@ all: install develop build builddoc check pylint test
 .PHONY: clobber
 clobber: clean
 	@echo "makefile: Removing everything for a fresh start"
-	rm -f *.done pylint.log flake8.log epydoc.log test_*.log $(moftab_files) $(dist_files) pywbem/*,cover wbemcli.log
-	rm -Rf $(doc_build_dir) .tox $(coverage_html_dir)
+	-$(call RM_FUNC,*.done epydoc.log $(moftab_files) $(dist_files) pywbem/*cover wbemcli.log)
+	-$(call RMDIR_FUNC,$(doc_build_dir) .tox $(coverage_html_dir))
 	@echo "makefile: Done removing everything for a fresh start"
 	@echo "makefile: Target $@ done."
 
@@ -427,11 +504,10 @@ clobber: clean
 .PHONY: clean
 clean:
 	@echo "makefile: Removing temporary build products"
-	$(FIND) . -name '*.pyc' -delete
-	$(FIND) . -name __pycache__ | xargs -n 1 rm -Rf
-	rm -Rf tmp_ tmp_*
-	rm -f MANIFEST parser.out .coverage $(package_name)/parser.out $(test_tmp_file)
-	rm -Rf build tmp_install testtmp tests/testtmp .cache $(package_name).egg-info .eggs
+	-$(call RM_R_FUNC,*.pyc)
+	-$(call RMDIR_R_FUNC,__pycache__)
+	-$(call RM_FUNC,MANIFEST parser.out .coverage $(package_name)/parser.out)
+	-$(call RMDIR_FUNC,build .cache $(package_name).egg-info .eggs)
 	@echo "makefile: Done removing temporary build products"
 	@echo "makefile: Target $@ done."
 
@@ -447,22 +523,22 @@ html: $(doc_build_dir)/html/docs/index.html
 	@echo "makefile: Target $@ done."
 
 $(doc_build_dir)/html/docs/index.html: makefile $(doc_utility_help_files) $(doc_dependent_files)
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the documentation as HTML pages"
-	rm -f $@
-	PYTHONPATH=. $(doc_cmd) -b html $(doc_opts) $(doc_build_dir)/html
+	-$(call RM_FUNC,$@)
+	$(doc_cmd) -b html $(doc_opts) $(doc_build_dir)/html
 	@echo "makefile: Done creating the documentation as HTML pages; top level file: $@"
 endif
 
 .PHONY: pdf
 pdf: makefile $(doc_utility_help_files) $(doc_dependent_files)
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the documentation as PDF file"
-	rm -f $@
+	-$(call RM_FUNC,$@)
 	$(doc_cmd) -b latex $(doc_opts) $(doc_build_dir)/pdf
 	@echo "makefile: Running LaTeX files through pdflatex..."
 	$(MAKE) -C $(doc_build_dir)/pdf all-pdf
@@ -472,11 +548,11 @@ endif
 
 .PHONY: man
 man: makefile $(doc_utility_help_files) $(doc_dependent_files)
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the documentation as man pages"
-	rm -f $@
+	-$(call RM_FUNC,$@)
 	$(doc_cmd) -b man $(doc_opts) $(doc_build_dir)/man
 	@echo "makefile: Done creating the documentation as man pages in: $(doc_build_dir)/man/"
 	@echo "makefile: Target $@ done."
@@ -484,7 +560,7 @@ endif
 
 .PHONY: docchanges
 docchanges:
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the doc changes overview file"
@@ -496,7 +572,7 @@ endif
 
 .PHONY: doclinkcheck
 doclinkcheck:
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the doc link errors file"
@@ -508,7 +584,7 @@ endif
 
 .PHONY: doccoverage
 doccoverage:
-ifeq ($(python_mn_version),26)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Sphinx doc build for target $@ on Python $(python_version)" >&2
 else
 	@echo "makefile: Creating the doc coverage results file"
@@ -521,10 +597,14 @@ endif
 # they are already specified e.g. in 'package_data' in setup.py.
 # We generate the MANIFEST.in file automatically, to have a single point of
 # control (this makefile) for what gets into the distribution archive.
-MANIFEST.in: makefile
+MANIFEST.in: makefile $(dist_manifest_in_files)
 	@echo "makefile: Creating the manifest input file"
 	echo "# file GENERATED by makefile, do NOT edit" >$@
+ifeq ($(PLATFORM),Windows_native)
+	for %%f in ($(dist_manifest_in_files)) do (echo include %%f >>$@)
+else
 	echo "$(dist_manifest_in_files)" |xargs -n 1 echo include >>$@
+endif
 	@echo "makefile: Done creating the manifest input file: $@"
 
 # Distribution archives.
@@ -534,16 +614,16 @@ MANIFEST.in: makefile
 # which can lead to incorrect hashbangs in the pywbem scripts in wheel archives.
 $(bdist_file) $(sdist_file): _check_version setup.py MANIFEST.in $(dist_dependent_files) $(moftab_files)
 	@echo "makefile: Creating the distribution archive files"
-	rm -rf MANIFEST $(package_name).egg-info .eggs build
-	rm -f PKG-INFO
+	-$(call RM_FUNC,MANIFEST PKG-INFO)
+	-$(call RMDIR_FUNC,build $(package_name).egg-info-INFO .eggs)
 	$(PYTHON_CMD) setup.py sdist -d $(dist_dir) bdist_wheel -d $(dist_dir) --universal
-	cp -r $(package_name).egg-info/PKG-INFO .
+	$(call CP_FUNC,$(package_name).egg-info/PKG-INFO,.)
 	@echo "makefile: Done creating the distribution archive files: $(bdist_file) $(sdist_file)"
 
 # Note: The mof*tab files need to be removed in order to rebuild them (make rules vs. ply rules)
 $(moftab_files): install.done $(moftab_dependent_files) build_moftab.py
 	@echo "makefile: Creating the LEX/YACC table modules"
-	rm -f $(package_name)/mofparsetab.py* $(package_name)/moflextab.py*
+	-$(call RM_FUNC,$(package_name)/mofparsetab.py* $(package_name)/moflextab.py*)
 	$(PYTHON_CMD) -c "from pywbem import mof_compiler; mof_compiler._build(verbose=True)"
 	@echo "makefile: Done creating the LEX/YACC table modules: $(moftab_files)"
 
@@ -558,57 +638,53 @@ $(moftab_files): install.done $(moftab_dependent_files) build_moftab.py
 # * 32 on usage error
 # Status 1 to 16 will be bit-ORed.
 # The make command checks for statuses: 1,2,32
-pylint.log: makefile $(pylint_rc_file) $(py_src_files)
-ifeq ($(python_mn_version),26)
+pylint.done: makefile $(pylint_rc_file) $(py_src_files)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Pylint on Python $(python_version)" >&2
 else
 	@echo "makefile: Running Pylint"
-	rm -f pylint.log
+	-$(call RM_FUNC,$@)
 	pylint --version
-	-bash -c 'set -o pipefail; PYTHONPATH=. pylint --rcfile=$(pylint_rc_file) $(py_src_files) 2>&1 |tee pylint.tmp.log; rc=$$?; if (($$rc >= 32 || $$rc & 0x03)); then exit $$rc; fi'
-	mv -f pylint.tmp.log pylint.log
-	@echo "makefile: Done running Pylint; Log file: $@"
+	-pylint --rcfile=$(pylint_rc_file) $(py_src_files)
+	echo "done" >$@
+	@echo "makefile: Done running Pylint"
 endif
 
-flake8.log: makefile $(flake8_rc_file) $(py_src_files)
-ifeq ($(python_mn_version),26)
+flake8.done: makefile $(flake8_rc_file) $(py_src_files)
+ifeq ($(python_mn_version),2.6)
 	@echo "makefile: Warning: Skipping Flake8 on Python $(python_version)" >&2
 else
 	@echo "makefile: Running Flake8"
-	rm -f flake8.log
+	-$(call RM_FUNC,$@)
 	flake8 --version
-	bash -c "set -o pipefail; PYTHONPATH=. flake8 --statistics --config=$(flake8_rc_file) --filename="*" $(py_src_files) 2>&1 |tee flake8.tmp.log"
-	mv -f flake8.tmp.log flake8.log
-	@echo "makefile: Done running Flake8; Log file: $@"
+	flake8 --statistics --config=$(flake8_rc_file) --filename='*' $(py_src_files)
+	echo "done" >$@
+	@echo "makefile: Done running Flake8"
 endif
 
-safety.log: makefile minimum-constraints.txt
+safety.done: makefile minimum-constraints.txt
 	@echo "makefile: Running pyup.io safety check"
-	rm -f $@
-	-bash -c "set -o pipefail; safety check -r minimum-constraints.txt --full-report |tee $@.tmp"
-	mv -f $@.tmp $@
-	@echo "makefile: Done running pyup.io safety check; Log file: $@"
+	-$(call RM_FUNC,$@)
+	-safety check -r minimum-constraints.txt --full-report
+	echo "done" >$@
+	@echo "makefile: Done running pyup.io safety check"
 
 .PHONY: test
-test: makefile $(package_name)/*.py $(mock_package_name)/*.py $(test_src_files) $(test_yaml_files) coveragerc
+test: $(moftab_files)
 	@echo "makefile: Running unit and function tests"
-	rm -f $(test_log_file)
-	bash -c "set -o pipefail; PYTHONWARNINGS=$(pytest_warnings) py.test --color=yes --cov $(package_name) --cov $(mock_package_name) $(coverage_report) --cov-config coveragerc $(pytest_opts) tests/unittest tests/functiontest -s 2>&1 |tee $(test_tmp_file)"
-	mv -f $(test_tmp_file) $(test_log_file)
-	@echo "makefile: Done running tests; Log file: $(test_log_file)"
+	py.test --color=yes --cov $(package_name) --cov $(mock_package_name) $(coverage_report) --cov-config coveragerc $(pytest_warnings_opts) $(pytest_opts) tests/unittest tests/functiontest -s
+	@echo "makefile: Done running tests"
 
 .PHONY: end2end
-end2end:
+end2end: $(moftab_files)
 	@echo "makefile: Running end2end tests"
-	rm -f $(test_end2end_log_file)
-	bash -c "set -o pipefail; PYTHONWARNINGS=$(pytest_end2end_warnings) py.test --color=yes $(pytest_end2end_opts) tests/end2endtest -s 2>&1 |tee $(test_end2end_tmp_file)"
-	mv -f $(test_end2end_tmp_file) $(test_end2end_log_file)
-	@echo "makefile: Done running end2end tests; Log file: $(test_end2end_log_file)"
+	py.test --color=yes $(pytest_end2end_warnings_opts) $(pytest_end2end_opts) tests/end2endtest -s
+	@echo "makefile: Done running end2end tests"
 
 $(doc_conf_dir)/wbemcli.help.txt: wbemcli wbemcli.py
 	@echo "makefile: Creating wbemcli script help message file"
-ifeq ($(PLATFORM),Windows)
-	cmd /d /c wbemcli.bat --help >$@
+ifeq ($(PLATFORM),Windows_native)
+	wbemcli.bat --help >$@
 else
 	./wbemcli --help >$@
 endif
@@ -616,8 +692,8 @@ endif
 
 $(doc_conf_dir)/mof_compiler.help.txt: mof_compiler $(package_name)/mof_compiler.py
 	@echo "makefile: Creating mof_compiler script help message file"
-ifeq ($(PLATFORM),Windows)
-	cmd /d /c mof_compiler.bat --help >$@
+ifeq ($(PLATFORM),Windows_native)
+	mof_compiler.bat --help >$@
 else
 	./mof_compiler --help >$@
 endif

--- a/pywbem_os_setup.bat
+++ b/pywbem_os_setup.bat
@@ -1,3 +1,14 @@
+@rem ---------------------------------------------------------------------------
+@rem Script that installs OS-level prerequisites for pywbem on native Windows
+@rem
+@rem Prerequisite commands for running this script:
+@rem     python (This script uses the active Python environment, virtual Python
+@rem       environments are supported)
+@rem     pip (with support for download subcommand, is installed by makefile)
+@rem     choco (Chocolatey package manager, from https://chocolatey.org)
+@rem     tar
+@rem     chmod
+
 @setlocal enableextensions
 @echo off
 set errorlevel=
@@ -49,18 +60,64 @@ echo Installing Swig ...
 
 set _CMD=choco install -y swig
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" echo Warning: choco returned rc=%_RC%
 
 set _CMD=where swig
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
 
 echo Done installing Swig
 :done_swig
+
+where curl >nul
+if %errorlevel%==0 (
+    echo Curl is already installed ... skipping
+    goto done_curl
+)
+
+echo Installing Curl ...
+
+set _CMD=choco install -y curl
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" echo Warning: choco returned rc=%_RC%
+
+set _CMD=where curl
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" goto error1
+
+echo Done installing Curl
+:done_curl
+
+where grep >nul
+if %errorlevel%==0 (
+    echo Grep is already installed ... skipping
+    goto done_grep
+)
+
+echo Installing Grep ...
+
+set _CMD=choco install -y grep
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" echo Warning: choco returned rc=%_RC%
+
+set _CMD=where grep
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" goto error1
+
+echo Done installing Grep
+:done_grep
 
 :: The bit size of Win32/64OpenSSL must match the bit size of the Python env.
 set _WINOPENSSL_BITSIZE=%PYTHON_ARCH%
@@ -75,14 +132,14 @@ echo Installing %_WINOPENSSL_BASENAME% ...
 
 set _CMD=curl -o %_WINOPENSSL_BASENAME%.exe -sSL https://slproweb.com/download/%_WINOPENSSL_BASENAME%.exe
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
 
 :: If the web site does not know the file, it returns a HTML page showing an error, and curl succeeds downloading that
 set _CMD=grep "^<html>" %_WINOPENSSL_BASENAME%.exe
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if "%_RC%"=="0" (
     echo Error: The %_WINOPENSSL_BASENAME%.exe file does not exist on https://slproweb.com:
@@ -93,16 +150,21 @@ if "%_RC%"=="0" (
 )
 
 :: Downloaded files may not have the execution right.
+rem TODO: Find a way to set RX with the built-in icacls to remove dependency on chmod
+rem set _CMD=icacls %_WINOPENSSL_BASENAME%.exe /grant "*S-1-1-0":F
 set _CMD=chmod 755 %_WINOPENSSL_BASENAME%.exe
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
+
+echo ACL permissions of %_WINOPENSSL_BASENAME%.exe:
+icacls %_WINOPENSSL_BASENAME%.exe
 
 :: The installer has a GUI which is suppressed by /silent /verysilent
 set _CMD=%_WINOPENSSL_BASENAME%.exe /silent /verysilent /suppressmsgboxes /dir="%_WINOPENSSL_INSTALL_DIR%"
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
 
@@ -120,13 +182,13 @@ echo Installing Python package M2Crypto version %_M2CRYPTO_VERSION% ...
 
 set _CMD=pip download M2Crypto==%_M2CRYPTO_VERSION%
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
 
 set _CMD=tar -xz -f M2Crypto-%_M2CRYPTO_VERSION%.tar.gz
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error1
 
@@ -134,13 +196,13 @@ pushd M2Crypto-%_M2CRYPTO_VERSION%
 
 set _CMD=python setup.py build --openssl=%_WINOPENSSL_INSTALL_DIR%
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error2
 
 set _CMD=python setup.py bdist_wheel
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error2
 
@@ -159,7 +221,7 @@ echo %_M2CRYPTO_WHEEL%
 
 set _CMD=pip install %_M2CRYPTO_WHEEL%
 echo %_CMD%
-%_CMD%
+call %_CMD%
 set _RC=%errorlevel%
 if not "%_RC%"=="0" goto error2
 
@@ -177,7 +239,31 @@ goto end
 
 echo Installing OS-level prerequisite packages for develop on platform Windows ...
 
+where xmllint >nul
+if %errorlevel%==0 (
+    echo xmllint is already installed ... skipping
+    goto done_xmllint
+)
+
+echo Installing xmllint ...
+
+set _CMD=choco install -y xsltproc
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" echo Warning: choco returned rc=%_RC%
+
+set _CMD=where xmllint
+echo %_CMD%
+call %_CMD%
+set _RC=%errorlevel%
+if not "%_RC%"=="0" goto error1
+
+echo Done installing xmllint
+:done_xmllint
+
 echo Warning: Package 'libxml2' must be installed manually.
+
 goto end
 
 :error2

--- a/pywbem_os_setup.bat
+++ b/pywbem_os_setup.bat
@@ -37,8 +37,8 @@ goto end
 
 echo Installing OS-level prerequisite packages for install on platform Windows ...
 
-if not exist ..\tmp_pywbem_os_setup mkdir ..\tmp_pywbem_os_setup
-pushd ..\tmp_pywbem_os_setup
+if not exist tmp_pywbem_os_setup mkdir tmp_pywbem_os_setup
+pushd tmp_pywbem_os_setup
 
 if not %PYTHON_M_VERSION%==2 goto install_no_py2
 
@@ -59,6 +59,8 @@ echo Installing Swig ...
 :: TODO: Remove the circumvention once it works in pywbem_os_setup.bat.
 
 set _CMD=choco install -y swig
+:: The following test ensures that this script still works when downloaded standalone.
+if exist ..\tools\retry.bat set _CMD=..\tools\retry.bat %_CMD%
 echo %_CMD%
 call %_CMD%
 set _RC=%errorlevel%
@@ -82,6 +84,8 @@ if %errorlevel%==0 (
 echo Installing Curl ...
 
 set _CMD=choco install -y curl
+:: The following test ensures that this script still works when downloaded standalone.
+if exist ..\tools\retry.bat set _CMD=..\tools\retry.bat %_CMD%
 echo %_CMD%
 call %_CMD%
 set _RC=%errorlevel%
@@ -105,6 +109,8 @@ if %errorlevel%==0 (
 echo Installing Grep ...
 
 set _CMD=choco install -y grep
+:: The following test ensures that this script still works when downloaded standalone.
+if exist ..\tools\retry.bat set _CMD=..\tools\retry.bat %_CMD%
 echo %_CMD%
 call %_CMD%
 set _RC=%errorlevel%
@@ -248,6 +254,8 @@ if %errorlevel%==0 (
 echo Installing xmllint ...
 
 set _CMD=choco install -y xsltproc
+:: The following test ensures that this script still works when downloaded standalone.
+if exist tools\retry.bat set _CMD=tools\retry.bat %_CMD%
 echo %_CMD%
 call %_CMD%
 set _RC=%errorlevel%
@@ -262,8 +270,6 @@ if not "%_RC%"=="0" goto error1
 echo Done installing xmllint
 :done_xmllint
 
-echo Warning: Package 'libxml2' must be installed manually.
-
 goto end
 
 :error2
@@ -275,5 +281,6 @@ echo Error: Command returned rc=%_RC%
 exit /b %_RC%
 
 :end
+rmdir /q /s tmp_pywbem_os_setup
 endlocal
 exit /b 0

--- a/pywbem_os_setup.sh
+++ b/pywbem_os_setup.sh
@@ -19,6 +19,16 @@
 # Author: Andreas Maier <maiera@de.ibm.com>
 #
 
+# Script that installs OS-level prerequisites for pywbem on Linux, MacOS, and
+# Windows with UNIX-like environments.
+#
+# Prerequisite commands for running this script:
+#     python (This script uses the active Python environment, virtual Python
+#       environments are supported)
+#     pip (with support for download subcommand, is installed by makefile)
+#     uname
+#     The package installer for your OS (yum, dnf, apt, zypper, brew)
+
 arg1="${1:-install}"
 arg2="${2:-}"
 myname=$(basename $0)

--- a/tests/unittest/utils/cmd_line_test_utils.py
+++ b/tests/unittest/utils/cmd_line_test_utils.py
@@ -259,13 +259,14 @@ def execute_wbemcli(args, verbose=False):
         if not isinstance(arg, six.text_type):
             arg = arg.decode('utf-8')
         cmd_args.append(arg)
+    cmd_args = u' '.join(cmd_args)
 
     if verbose:
-        print('CMD_ARGS %s' % cmd_args)
+        print('Calling with shell: %r' % cmd_args)
 
     # Using universal_newlines=True reduces EOLs to
     # '\n' despite what other packages might add on.
-    proc = Popen(cmd_args, shell=False, stdout=PIPE, stderr=PIPE,
+    proc = Popen(cmd_args, shell=True, stdout=PIPE, stderr=PIPE,
                  universal_newlines=True)
     stdout_str, stderr_str = proc.communicate()
     rc = proc.returncode

--- a/tests/unittest/utils/validate.py
+++ b/tests/unittest/utils/validate.py
@@ -67,7 +67,7 @@ import os.path
 import re
 from subprocess import Popen, PIPE, STDOUT
 
-from pywbem._utils import _ensure_bytes
+from pywbem._utils import _ensure_bytes, _ensure_unicode
 
 # CIM-XML DTD file to validate against
 DTD_FILE = os.path.join('tests', 'dtd', 'DSP0203_2.3.1.dtd')
@@ -125,10 +125,10 @@ def validate_cim_xml(cim_xml_str, root_elem_name=None):
     p.stdin.write(cim_xml_str)
     p.stdin.close()
 
-    output = '\n'.join(p.stdout.readlines())
-
     status = p.wait()
     if status != 0:
+        out_lines = p.stdout.readlines()
+        output = _ensure_unicode(b'\n'.join(out_lines))
         raise CIMXMLValidationError(output)
 
 

--- a/tools/package_version.py
+++ b/tools/package_version.py
@@ -1,0 +1,17 @@
+#!/usr/bin/env python
+
+import sys
+try:
+    from pbr.version import VersionInfo
+except ImportError:
+    # Version cannot be determined
+    sys.exit(0)
+
+if len(sys.argv) != 2:
+    print("Error: Incorrect number of command line arguments")
+    print("Usage: {0}  package_name".format(sys.argv[0]))
+    sys.exit(2)
+
+pkg_name = sys.argv[1]
+pkg_version = VersionInfo(pkg_name).release_string()
+print(pkg_version)

--- a/tools/python_bitsize.py
+++ b/tools/python_bitsize.py
@@ -1,0 +1,7 @@
+#!/usr/bin/env python
+
+import ctypes
+
+pointer_bits = ctypes.sizeof(ctypes.c_void_p) * 8
+
+print("Bit size of Python environment: {0} bit".format(pointer_bits))

--- a/tools/python_version.py
+++ b/tools/python_version.py
@@ -1,0 +1,20 @@
+#!/usr/bin/env python
+
+import sys
+
+if len(sys.argv) != 2:
+    print("Error: Incorrect number of command line arguments")
+    print("Usage: {0}  version_level".format(sys.argv[0]))
+    print("Where version_level is the number of components in the "
+          "version string from 1 to 3, e.g. 3 means Python 2.7.11")
+
+    sys.exit(2)
+
+version_level = sys.argv[1]
+if version_level not in ('1', '2', '3'):
+    print("Error: Invalid version level; must be 1, 2, or 3")
+    sys.exit(2)
+
+version_level = int(version_level)
+py_version = '.'.join([str(v) for v in sys.version_info[0:version_level]])
+print(py_version)

--- a/tools/retry.bat
+++ b/tools/retry.bat
@@ -1,0 +1,24 @@
+@echo off
+rem Retry a command for a number of times with sleep time in between retries.
+
+rem Configuration area:
+set max_attempts=3
+set sleep_time=30
+
+set attempts=0
+:try
+%*
+set cmd_rc=%errorlevel%
+if %cmd_rc% == 0 goto :done
+set /a attempts=%attempts%+1
+if %attempts% == %max_attempts% goto fail
+set /a next_attempt=%attempts%+1
+timeout %sleep_time%
+@echo Info: Command "%*" failed with exit code %cmd_rc% - now trying attempt %next_attempt% of %max_attempts%
+goto try
+
+:fail
+@echo Error: Giving up on command "%*" after exceeding the maximum of %max_attempts% attempts
+
+:done
+exit /b %cmd_rc%

--- a/tox.ini
+++ b/tox.ini
@@ -27,9 +27,11 @@ commands =
     env
     tox --version
     python --version
-    pip list
+    make platform env
     rm -f install_basic.done install_pywbem.done install.done develop.done
+    pip list
     make install
+    pip list
     make develop
     pip list
     make test

